DESIGN.md

```diff
@@ -0,0 +1,126 @@
+    - Created
+    - In Process
+    - Exited
+    - Error
+- **Job**
+    - Command (string):
+    Command holds a valid shell command that is run on the underlying server shell.   
+    - JID (integer):
+    JID holds an integer representing job ID. 
+    - CreatedTime (string):
+    Timestamp that represents when the process was launched on the server
+    - JobExecutionTime (string):
+    Timestamp that represents how long the underlying process has been running
+    - FinishedTime (string):
+    Timestamp that represents when a Job finished
+    - Status (status):
+    Instance of status enumeration. Represents state of job
+    - Output (string):
+    Output contains the output of the job at time it is retrieved
+- **JobList**
+    - Jobs ([]Job):
+    A list of Job obejcts as defined above in this list
+- **DeleteJobOutput**
+    - Job (Job):
+    Job object as defined above in this list
+    - Exit (integer):
+    Exit code as returned from the underlying process
+    - State (status):
+    Instance of status enumeration representing the state of the job at time of deletion
+
+### Data flow
+Each handler will receive the endpoint input as well as a context object from which the client certificate and trust chain can be extracted. After the Authorization routine is applied to the CA certificate found in the client certificate trust chain, the handler will process the input and return the output.
+
+![flowchart for CreateJob API call](doc/data_flow.png)
+
+## Controller
+The backend will leverage the CommandContext type from the os/exec package. This will provide convenient facilities for cancelling running jobs and garnering job output. Additionally, it will handle passing through environment variables for us.
+
+A single datatype, JobCtl, will be defined to contain the following peices of data:
+
+- The os/exec CommandContext instance
+- The cancel function derived from the context passed into the CommandContext instance
+- The "Job" instance as defined in the API datatypes.
+    
+A map will be used to cache JobCtl objects. The JobCtl objects will be mapped to their JID. A mutex will be used to synchronize concurrent access to the cache. Methods will be defined to retrieve API Job objects, create new JobCtls, delete existing JobCtls, and fetch output from a JobCtl.
+
+## Client
+A command line client will be created from the compiled client library stub code. The command line client will use arguments for RPC calls, arguments, and certificates. Below is a list of important arguments:
+
+- client_cert: filepath to the certificate used by the client
+- client_key: filepath to the private key used by the client
+- server_ca: filepath to the CA cert that signs the server cert
+- op: requested remote procedure. One of {CreateJob, ListJobs, GetJob, DeleteJob}.
+- arg: Optional argument for the remote procedure. Argument could be a JID or a shell command
+
+Below are some examples of CLI invocations:
+
+```sh
```

Comment:
the cli doesn't offer great ux, but I am fine if you want to keep it as is

Reply:
In order to improve the UX I will use environment variables to set the certificates and keys.